File: Java/programmers/etc/level3/N으로표현.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class N으로표현 {
    public int solution(int N, int number) {
        if (N == number) return 1;
        
        List<Set<Integer>> set = new ArrayList<>();

        for (int i=0; i<8; i++) {
            set.add(new HashSet<>());
            set.get(i).add(Integer.parseInt(String.valueOf(N).repeat(i+1)));
        }
        
        int answer = -1;
        for (int i=1; i<8; i++) {
            for (int j=0; j<i; j++) {
                for (Integer op1 : set.get(j)) {
                    for (Integer op2 : set.get(i-j-1)) {
                        set.get(i).add(op1+op2);
                        set.get(i).add(op1-op2);
                        set.get(i).add(op1*op2);
                        if (op2 != 0) set.get(i).add(op1/op2);
                    }
                }
            }
            if (set.get(i).contains(number)) {
                answer = i+1;
                break;
            }
        }
        
        return answer;
    }
}

File: Python/programmers/etc/level3/N으로_표현.py
def solution(N, number):
    if N == number:
        return 1
    s = [set() for _ in range(8)]
    for i, x in enumerate(s):
        x.add(int(str(N)*(i+1)))
        
    for i in range(1, 8):
        for j in range(i):
            for op1 in s[j]:
                for op2 in s[i-j-1]:
                    s[i].add(op1+op2)
                    s[i].add(op1-op2)
                    s[i].add(op1*op2)
                    if op2 != 0:
                        s[i].add(op1//op2)
                        
        if number in s[i]:
            answer = i+1
            break
    else:
        answer = -1
        
    return answer


N, number = 5, 12
print(solution(N, number)) # 4

N, number = 2, 11
print(solution(N, number)) # 3
